Sources/NIO/SelectorGeneric.swift

```diff
@@ -0,0 +1,376 @@
+        if var this = self {
+            return try withUnsafePointer(to: &this) { x in
+                try body(x)
+            }
+        } else {
+            return try body(nil)
+        }
+    }
+}
+
+extension timespec {
+    init(timeAmount amount: TimeAmount) {
+        let nsecPerSec: Int64 = 1_000_000_000
+        let ns = amount.nanoseconds
+        let sec = ns / nsecPerSec
+        self = timespec(tv_sec: Int(sec), tv_nsec: Int(ns - sec * nsecPerSec))
+    }
+}
+
+/// Represents IO events NIO might be interested in. `SelectorEventSet` is used for two purposes:
+///  1. To express interest in a given event set and
+///  2. for notifications about an IO event set that has occured.
+///
+/// For example, if you were interested in reading and writing data from/to a socket and also obviously if the socket
+/// receives a connection reset, express interest with `[.read, .write, .reset]`.
+/// If then suddenly the socket becomes both readable and writable, the eventing mechanism will tell you about that
+/// fact using `[.read, .write]`.
+struct SelectorEventSet: OptionSet, Equatable {
+
+    typealias RawValue = UInt8
+
+    let rawValue: RawValue
+
+    /// It's impossible to actually register for no events, therefore `_none` should only be used to bootstrap a set
+    /// of flags or to compare against spurious wakeups.
+    static let _none = SelectorEventSet([])
+
+    /// Connection reset or other errors.
+    static let reset = SelectorEventSet(rawValue: 1 << 0)
+
+    /// EOF at the read/input end of a `Selectable`.
+    static let readEOF = SelectorEventSet(rawValue: 1 << 1)
+
+    /// Interest in/availability of data to be read
+    static let read = SelectorEventSet(rawValue: 1 << 2)
+
+    /// Interest in/availability of data to be written
+    static let write = SelectorEventSet(rawValue: 1 << 3)
+
+    /// EOF at the write/output end of a `Selectable`.
+    ///
+    /// - note: This is rarely used because in many cases, there is no signal that this happened.
+    static let writeEOF = SelectorEventSet(rawValue: 1 << 4)
+
+    init(rawValue: SelectorEventSet.RawValue) {
+        self.rawValue = rawValue
+    }
+}
+
+internal let isEarlyEOFDeliveryWorkingOnThisOS: Bool = {
+    #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+    return false // rdar://53656794 , once fixed we need to do an OS version check here.
+    #else
+    return true
+    #endif
+}()
+
+/// This protocol defines the methods that are expected to be found on
+/// `Selector`. While defined as a protocol there is no expectation that any
+/// object other than `Selector` will implement this protocol: instead, this
+/// protocol acts as a reference for what new supported selector backends
+/// must implement.
+protocol _SelectorBackendProtocol {
+    associatedtype R : Registration
+    func _init() throws
+    func _deinit()
+    func _register<S: Selectable>(selectable: S, fd: Int, interested: SelectorEventSet) throws
```

Comment:
typically `_functionName` means that a function has a higher access control level than it should have (for example `public func _foo` means "this should really be internal but can't be for some reason").

In this case I'm not sure if they should be `_`'d. I think we should probably call them `registerImpl` or `register0`, the `functionName0` being a Netty-ism...

Said that, I can live with `_functionName` too. @Lukasa wdyt?

Reply:
Just let me know what you prefer - probably would go for the Netty-ism in that case if you don't like`_Name` as it is used in other places of the code (I've tried to decipher the `name0` as never encountered that, so thanks for pointing out origin).

Reply:
`_Name` is only used for "is public, should be internal" or "is internal, should be private (usually to allow @usableFromInline)" in NIO so far. Most other NIO code uses the `name0`. So I too think that we should follow this

Reply:
Done in latest.